Toolkit/ArcGISToolkit/TemplatePickerViewController.swift

```diff
@@ -0,0 +1,303 @@
+// Copyright 2016 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ArcGIS
+
+/// An object that encapsulates information related to a feature template
+public class FeatureTemplateInfo{
+    /// The feature layer that the template is from
+    public let featureLayer: AGSFeatureLayer
+    /// The feature table that the template is from
+    public let featureTable: AGSArcGISFeatureTable
+    /// The feature template
+    public let featureTemplate: AGSFeatureTemplate
+    /// The swatch for the feature template
+    public var swatch: UIImage?
+    
+    fileprivate init(featureLayer: AGSFeatureLayer, featureTable: AGSArcGISFeatureTable, featureTemplate: AGSFeatureTemplate, swatch: UIImage? = nil){
+        self.featureLayer = featureLayer
+        self.featureTable = featureTable
+        self.featureTemplate = featureTemplate
+        self.swatch = swatch
+    }
+}
+
+/// The protocol you implement to respond as the user interacts with the feature templates
+/// view controller.
+public protocol TemplatePickerViewControllerDelegate: AnyObject {
+    /// Tells the delegate that the user has cancelled selecting a template.
+    ///
+    /// - Parameter templatePickerViewController: The current template picker view controller.
+    func templatePickerViewControllerDidCancel(_ templatePickerViewController: TemplatePickerViewController)
+    /// Tells the delegate that the user has selected a feature template.
+    ///
+    /// - Parameters:
+    ///   - templatePickerViewController: The current template picker view controller.
+    ///   - featureTemplateInfo: The selected feature template.
+    func templatePickerViewController(_ templatePickerViewController: TemplatePickerViewController, didSelect featureTemplateInfo: FeatureTemplateInfo)
+}
+
+/// A view controller that is useful for showing the user a list of feature templates
+/// and allowing them to choose one.
+/// This view controller is meant to be embedded in a navigation controller.
+public class TemplatePickerViewController: TableViewController {
+    
+    /// The map which this view controller will display the feature templates from
+    public private(set) var map: AGSMap?
```

Comment:
I just noticed that the map is assigned in the initializer and never mutated. Could it be made immutable?

```swift
public let map: AGSMap

public init(map: AGSMap) {
    self.map = map
    super.init(nibName: nil, bundle: nil)
}
```

Reply:
good call - i will fix